dali/python/nvidia/dali/_utils/eager_util.py

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import nvidia.dali.tensors as _tensors
+import nvidia.dali.types as _types
+from nvidia.dali.external_source import _prep_data_for_feed_input
+
+
+def _transform_data_to_tensorlist(data, batch_size, layout=None, device_id=None):
+    data = _prep_data_for_feed_input(data, batch_size, layout, device_id)
+
+    if isinstance(data, list):
+        if isinstance(data[0], _tensors.TensorGPU):
+            data = _tensors.TensorListGPU(data, layout or "")
+        else:
+            data = _tensors.TensorListCPU(data, layout or "")
+
+    return data
+
+
+class _Classification:
+    """Classification of data's device and if it is a batch.
+
+    Based on data type determines if data should be treated as a batch and with which device.
+    If the type can be recognized as a batch without being falsely categorized as such, it is.
+    This includes lists of supported tensor-like objects e.g. numpy arrays (the only list not
+    treated as a batch is a list of objects of primitive types), :class:`DataNodeDebug` and
+    TensorLists.
+
+    Args:
+        data: Data to be classified.
+        type_name (str): Representation of argument type (input or keyward).
+        arg_batch_size (int): Size of batch to which constant value should be scaled up. Only
+            applicable for argument inputs that are array types (e.g. numpy array). If -1 does not
+            modify the data.
+    """
+
+    def __init__(self, data, type_name, arg_batch_size=-1):
+        self.is_batch, self.device, self.data = self._classify_data(data, type_name, arg_batch_size)
+
+    @staticmethod
+    def _classify_data(data, type_name, arg_batch_size):
+        from nvidia.dali._debug_mode import DataNodeDebug
+        """Returns tuple (is_batch, device, unpacked data). """
+
+        def is_primitive_type(x):
+            return isinstance(x, (int, float, bool, str))
+
+        def classify_array_type(data):
+            if _types._is_numpy_array(data):
+                device = 'cpu'
+            elif _types._is_torch_tensor(data):
+                device = 'gpu' if data.is_cuda else 'cpu'
+                
+                # Data needs to be on CPU for batch argument inputs.
+                if device == 'gpu' and arg_batch_size > 0:
+                    data = data.cpu().numpy()
+            elif _types._is_mxnet_array(data):
+                import mxnet as mx
+
+                device = 'gpu' if 'gpu' in str(data.context) else 'cpu'
+
+                # For batch copy to CPU.
+                if device == 'gpu' and arg_batch_size > 0:
+                    data = data.copyto(mx.cpu())
+            else:
+                raise RuntimeError(f"Unsupported array type '{type(data)}'.")
+            
+            data = _types._preprocess_constant_array_type(data)[0]
```

Comment:
I don't get where this `[0]` comes from here. Aren't we taking slice of the data now?

Reply:
`_types._preprocess_constant_array_type` returned tuple and we took the first element of that tuple - data. I modified it so it returns just the data now.